scheduler/src/cook/scheduler/scheduler.clj

```diff
@@ -767,11 +770,54 @@
                                        {:job-uuid->reserved-host (apply dissoc job-uuid->reserved-host matched-job-uuids)
                                         :launched-job-uuids (into matched-job-uuids launched-job-uuids)})))
 
+(defn distribute-task-requests-to-compute-clusters
+  "Given a collection of un-matched task requests and a collection of
+  compute clusters, distributes the task requests amongst the compute
+  clusters, using a hash of the pending job's uuid. Returns a
+  compute-cluster->task-request map. That is the API any future
+  improvements need to stick to."
+  [task-requests compute-clusters]
+  (group-by (fn [{:keys [job]}]
+              (nth compute-clusters
+                   (-> job :job/uuid hash (mod (count compute-clusters)))))
+            task-requests))
+
+(defn trigger-autoscaling!
+  "Autoscales the given pool to satisfy the given match failures (i.e. pending jobs), if:
+  - There is at least one failure
+  - There is at least one compute cluster configured to do autoscaling"
+  [failures pool-name compute-clusters]
+  (try
+    (let [task-requests (map #(.. (first %) (getRequest)) failures)
```

Comment:
I"m not sure if the logic belongs here or above (see comments in pending-offers, but:

1. We should probably autoscale at a longer granularity than every offer cycle. If it takes 3 minutes to launch a node, there's not much point autoscaling much more rapidly than that. The system (k8s) can't respond that fast. Thus, autoscale maybe every 20 offer cycles. this lessens the impact of waiting for our synthetic pods to be launched by k8s (See comment in pending-offers). (I prefer to go by offer cycle, that way the overhead is constant  fraction of the work done.) If offer cycle is every 10 seconds, 20 cycles is 3 minutes, comparable to node launch time.

2. We don't generate a synthetic pod for each unmatched job. We give ourselves some hysteresis. E.g., We only generate synthetic pods for the first half of the outstanding queue. E.g., min(max-pods-outstanding, count(failures)/2),

3. If autoscaling cycle is ~3 minutes (the time I assume to launch a node), then we can autoscale by a max of 200 pods every 3 minutes. If we had 5k jobs suddenly added, it would take 25 cycles, or about 120 minutes to scale up to hit it.  We might want to consider increasing maximum of max-pods-oustanding to 500 or 1000.

Reply:
@rmanyari - Thoughts on this?